gadopt/level_set_tools.py

```diff
@@ -0,0 +1,239 @@
+            self.mesh, "CG", func_space.finat_element.degree
+        )
+        self.level_set_grad_proj = Function(
+            self.func_space_lsgp,
+            name=f"Level-set gradient (projection) #{level_set.name()[-1]}",
+        )
+        self.proj_solver = self.gradient_L2_proj()
+
+        self.reini_params = reini_params
+        reini_fields = {
+            "level_set_grad": self.level_set_grad_proj,
+            "epsilon": reini_params["epsilon"],
+        }
+
+        ls_eq = ScalarAdvectionEquation(func_space, func_space)
+        reini_eq = ReinitialisationEquation(func_space, func_space)
+
+        if solver_params is None:
+            solver_params = {
+                "mat_type": "aij",
+                "ksp_type": "preonly",
+                "pc_type": "bjacobi",
+                "sub_pc_type": "ilu",
+            }
+
+        self.ls_ts = tstep_alg(
+            ls_eq,
+            self.level_set,
+            ls_fields,
+            tstep / subcycles,
+            solver_parameters=solver_params,
+        )
+        self.reini_ts = reini_params["tstep_alg"](
+            reini_eq,
+            self.level_set,
+            reini_fields,
+            reini_params["tstep"],
+            solver_parameters=solver_params,
+        )
+
+        self.subcycles = subcycles
+
+    def gradient_L2_proj(self):
+        test_function = TestFunction(self.func_space_lsgp)
+        trial_function = TrialFunction(self.func_space_lsgp)
+
+        mass_term = inner(test_function, trial_function) * dx(domain=self.mesh)
+        residual_element = self.level_set * div(test_function) * dx(domain=self.mesh)
+        residual_boundary = (
+            self.level_set
+            * dot(test_function, FacetNormal(self.mesh))
+            * ds(domain=self.mesh)
+        )
+        residual = -residual_element + residual_boundary
+        problem = LinearVariationalProblem(
+            mass_term, residual, self.level_set_grad_proj
+        )
+
+        solver_params = {
+            "mat_type": "aij",
+            "ksp_type": "preonly",
+            "pc_type": "lu",
+            "pc_factor_mat_solver_type": "mumps",
+        }
+        return LinearVariationalSolver(problem, solver_parameters=solver_params)
+
+    def update_level_set_gradient(self, *args):
+        self.proj_solver.solve()
+
+    def solve(self, step):
+        for subcycle in range(self.subcycles):
+            self.ls_ts.advance(0)
+
+            if step >= self.reini_params["frequency"]:
+                self.reini_ts.solution_old.assign(self.level_set)
+
+            if step % self.reini_params["frequency"] == 0:
+                for reini_step in range(self.reini_params["iterations"]):
+                    self.reini_ts.advance(
+                        0, update_forcings=self.update_level_set_gradient
+                    )
+
+                self.ls_ts.solution_old.assign(self.level_set)
+
+
+def sharp_interface(level_set, material_value, method):
```

Comment:
```suggestion
def sharp_interface(level_set, material_value, method):
    """docstring!"""
    sharp_interface_popped(level_set.copy(), material_value.copy(), method)
    
def sharp_interface_popped(level_set, material_value, method):
    """Construct UFL expression defining sharp interface by popping list of level_sets and material_values"""
```

Reply:
I think it's better to be safe with these user-facing functions, and not eat the arguments provided by the user. Copies are trivially cheap anyway